api/src/controllers/login.js

```diff
@@ -355,6 +515,47 @@ module.exports = {
       });
   },
 
+  getPasswordReset: (req, res, next) => {
+    return renderPasswordReset(req)
+      .then(body => {
+        res.setHeader(
+          'Link',
+          '</login/style.css>; rel=preload; as=style, '
+              + '</login/auth-utils.js>; rel=preload; as=script, '
+              + '</login/password-reset.js>; rel=preload; as=script'
+        );
+        res.send(body);
+      })
+      .catch(next);
+  },
+  resetPassword: async (req, res) => {
+    const limited = await rateLimitService.isLimited(req);
+    if (limited) {
+      return serverUtils.rateLimited(req, res);
+    }
+
+    try {
+      const { username, currentPassword, password } = req.body;
+      const validationResult = await passwordResetValidation(username, currentPassword, password);
+      if (!validationResult.isValid) {
+        return res.status(validationResult.status).json({
+          error: validationResult.error,
+          params: validationResult.params
+        });
+      }
+
+      const userDoc = await db.users.get(`org.couchdb.user:${username}`);
+      await updatePassword(userDoc, password);
```

Comment:
It's because of a race condition, when testing it out, in some instances  requesting session immediately after changing a password was giving back a 401. Hence I safe guarded it with the retry to avoid that.

Reply:
So, yeah, the `createSessionRetry` logic makes sense to me (and I guess already existed for token login).  But here is where we are actually _changing_ the password.  Seems like we should not be getting a `401` from the `db.users.put(updatedUser);` call.  If I understand things correctly, the user session that is actually used to make that update would be the api server's admin user (`COUCH_USER`) anyway since we have not yet created a session for the current user trying to login....